data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/PeerForwarder.java

```diff
@@ -85,7 +96,9 @@ public PeerForwarder(final PluginSetting pluginSetting, final PeerForwarderConfi
         }
 
         // Buffer of requests to be exported to the downstream of the local data-prepper
-        final List<Record<ExportTraceServiceRequest>> results = new ArrayList<>();
+        final List<Record<ExportTraceServiceRequest>> results = new LinkedList<>();
+
+        final List<CompletableFuture<Record>> futures = new LinkedList<>();
```

Comment:
There may be a misunderstanding regarding the intent of the 2 lists - I can try to clean this up to make it a bit more clear (naming, comments, etc.). 

* `results` - a list of records containing ExportTraceServiceRequests which are intended to be passed further down the pipeline to be processed on this host.
* `futures` - a list of CompletableFutures which represent calls made to other peers. If the call succeeds, the future returns null (might be better as an Optional), if the call fails, the future returns the ExportTraceServiceRequest so that _this_ Data Prepper instance can process it locally (this is paraphrased in the `processRequest` docblock).

As we iterate over each ETSR - if the request is meant to be handled locally (`client == null`), great, add it to the `results` list. If the request is meant to be handled by a peer (`client != null`), kick off a CompletableFuture that will either return nothing or something. After all requests have been made, go ahead and get the results of each future. Any failed forwardings will have a non-null record, at which point we can add it to `results` to process it locally too.

We definitely _could_ combine the lists into a single list containing only Futures. In this case, records meant to be processed locally would be wrapped in a Future which immediately returns the record back (doesn't make a peer call). But now we're just creating wrapper objects to cut out a list, which takes time and space. I think 2 lists are fine, it's just the implementation could be made a bit more clear.

Reply:
Thanks for clarifying!